File: apps/web/src/features/members/api/memberProfile.ts
import { apiGet, apiPut } from '@/lib/api/client';

export type RelationshipType =
  | 'parent'
  | 'child'
  | 'spouse'
  | 'sibling'
  | 'grandparent'
  | 'grandchild'
  | 'aunt'
  | 'uncle'
  | 'cousin'
  | 'niece'
  | 'nephew'
  | 'in_law'
  | 'friend'
  | 'colleague'
  | 'mentor'
  | 'mentee'
  | 'caregiver'
  | 'neighbor'
  | 'other';

export interface MemberProfile {
  relationship_type: RelationshipType | null;
  nickname: string | null;
  legacy_to_viewer: string | null;
  viewer_to_legacy: string | null;
  character_traits: string[] | null;
}

export interface MemberProfileUpdate {
  relationship_type?: RelationshipType | null;
  nickname?: string | null;
  legacy_to_viewer?: string | null;
  viewer_to_legacy?: string | null;
  character_traits?: string[];
}

export const RELATIONSHIP_TYPE_LABELS: Record<RelationshipType, string> = {
  parent: 'Parent',
  child: 'Child',
  spouse: 'Spouse',
  sibling: 'Sibling',
  grandparent: 'Grandparent',
  grandchild: 'Grandchild',
  aunt: 'Aunt',
  uncle: 'Uncle',
  cousin: 'Cousin',
  niece: 'Niece',
  nephew: 'Nephew',
  in_law: 'In-Law',
  friend: 'Friend',
  colleague: 'Colleague',
  mentor: 'Mentor',
  mentee: 'Mentee',
  caregiver: 'Caregiver',
  neighbor: 'Neighbor',
  other: 'Other',
};

export async function getMemberProfile(
  legacyId: string
): Promise<MemberProfile | null> {
  return apiGet<MemberProfile | null>(`/api/legacies/${legacyId}/profile`);
}

export async function updateMemberProfile(
  legacyId: string,
  data: MemberProfileUpdate
): Promise<MemberProfile> {
  return apiPut<MemberProfile>(`/api/legacies/${legacyId}/profile`, data);
}

File: services/core-api/app/schemas/member_profile.py
"""Pydantic schemas for member relationship profiles."""

from enum import Enum

from pydantic import BaseModel, Field


class RelationshipType(str, Enum):
    """Supported relationship types."""

    parent = "parent"
    child = "child"
    spouse = "spouse"
    sibling = "sibling"
    grandparent = "grandparent"
    grandchild = "grandchild"
    aunt = "aunt"
    uncle = "uncle"
    cousin = "cousin"
    niece = "niece"
    nephew = "nephew"
    in_law = "in_law"
    friend = "friend"
    colleague = "colleague"
    mentor = "mentor"
    mentee = "mentee"
    caregiver = "caregiver"
    neighbor = "neighbor"
    other = "other"


class GenderType(str, Enum):
    """Supported gender options."""

    male = "male"
    female = "female"
    non_binary = "non_binary"
    prefer_not_to_say = "prefer_not_to_say"


class MemberProfileUpdate(BaseModel):
    """Request to create or update a member's relationship profile."""

    relationship_type: RelationshipType | None = None
    nickname: str | None = Field(None, max_length=100)
    legacy_to_viewer: str | None = Field(None, max_length=1000)
    viewer_to_legacy: str | None = Field(None, max_length=1000)
    character_traits: list[str] | None = None


class MemberProfileResponse(BaseModel):
    """Response containing a member's relationship profile."""

    relationship_type: RelationshipType | None = None
    nickname: str | None = None
    legacy_to_viewer: str | None = None
    viewer_to_legacy: str | None = None
    character_traits: list[str] | None = None
